### DISCUSSION TITLE
Serialising large amounts of float values that don't need much accuracy

### DISCUSSION THREAD

**Discussion by silentdiverchris** [nlohmann/json#3733] [Q&A] [answered]
Hello there, I'm looking for a way to reduce the size of the json output where I am writing a large number of floats, I round them to 3dps before serialising but am getting results along the lines of 0.12300004567, which puzzled me, but then I read up and found info about how floats are handled, with the %g printf modifier with [std::numeric_limits<double>::max_digits10] and all that - which explains the mysterious 4567 appearing, the float-double conversion is adding the imprecision.

I can't change the float to double or decimal, and 3 decimals is more than enough accuracy for my purpose (the numbers represent rough probabilities, so are all between 0 and 1), so I'm mainly just trying to make the output json smaller by only writing 3 decimal places to it, which would drop the size of the output by well over a half so definitely worth the tweak.

C++ isn't my thing, and I'm working on an existing system so large or ambitious changes are not on the cards, and writing a custom serialiser as mentioned in the readme would be overkill/difficult etc, I'm just wondering if there is a tweak I can make in my particular implementation to just cut off floats written to the json at 3 decimals.

**Comment by falbrechtskirchinger** [comment, marked as answer]
Making a minor modification to the source that hard-codes a precision of 3 is pretty easy:

Find the `dump_float()` function in the serializer:
https://github.com/nlohmann/json/blob/307c053b9b250abc6e6d478a4336fd98592ae173/include/nlohmann/detail/output/serializer.hpp#L820-L826

Modify as follows:
```cpp
        auto* begin = number_buffer.data();
        //auto* end = ::nlohmann::detail::to_chars(begin, begin + number_buffer.size(), x);
        auto [end, ec] = std::to_chars(begin, begin + number_buffer.size(), x, std::chars_format::fixed, 3);

        o->write_characters(begin, static_cast<size_t>(end - begin));
```
Add `#include <charconv>` to the top.

This solution requires C++17. If you need it to work with C++11/C++14 you can use any other method to format the number `x` and write it to the output `o`.

**Comment by silentdiverchris** [comment]
Excellent, thank you, it seems the version of the library that's embroiled with this project is nowhere near the latest, and it's definitely not for me to start updating it, we're not using C++ 17 and are aiming for the absolute minimum change so - here's what I did, and it does the trick, since by definition, all my numbers are between 0.001 and 1.000.
```

char* begin = number_buffer.data();
char* end = ::nlohmann::detail::to_chars(begin, begin + number_buffer.size(), x);

// Tweak the end pointer to take the first 5 at most
if (end - begin > 5) end = begin + 5;

o->write_characters(begin, static_cast<size_t>(end - begin));

```

Thanks again for such a swift response !
